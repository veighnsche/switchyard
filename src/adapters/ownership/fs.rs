// Default OwnershipOracle implementation using OS metadata (Unix-only)

use crate::adapters::OwnershipOracle;
use crate::types::OwnershipInfo;
use crate::types::errors::{Error, ErrorKind, Result};
use crate::types::safepath::SafePath;

#[derive(Copy, Clone, Debug, Default)]
pub struct FsOwnershipOracle;

impl OwnershipOracle for FsOwnershipOracle {
    fn owner_of(&self, path: &SafePath) -> Result<OwnershipInfo> {
        #[cfg(unix)]
        {
            use std::os::unix::fs::MetadataExt;
            let md = std::fs::symlink_metadata(path.as_path()).map_err(|e| Error {
                kind: ErrorKind::Io,
                msg: format!("metadata: {}", e),
            })?;
            Ok(OwnershipInfo {
                uid: md.uid(),
                gid: md.gid(),
                pkg: String::new(),
            })
        }
        #[cfg(not(unix))]
        {
            Err(Error {
                kind: ErrorKind::Policy,
                msg: "OwnershipOracle not supported on this platform".into(),
            })
        }
    }
}
