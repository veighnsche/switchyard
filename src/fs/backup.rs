//! Backup and sidecar helpers for Switchyard filesystem operations.
//!
//! This module centralizes the backup payload and sidecar schema handling used
//! by symlink replacement and restore operations.

/// replace this file with src/fs/backup/{mod.rs,snapshot.rs,sidecar.rs,index.rs} â€” split monolith per zrefactor/fs_refactor_backup_restore.INSTRUCTIONS.md

mod snapshot;
mod sidecar;
mod index;

pub use snapshot::*;
pub use sidecar::*;
pub use index::*;
