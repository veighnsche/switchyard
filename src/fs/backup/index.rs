use std::{
    collections::HashSet,
    fs,
    path::{Path, PathBuf},
};

use super::sidecar::sidecar_path_for_backup;

/// Return (`backup_path_if_present`, `sidecar_path`) for the latest timestamped pair.
pub(crate) fn find_latest_backup_and_sidecar(
    target: &Path,
    tag: &str,
) -> Option<(Option<PathBuf>, PathBuf)> {
    let name = target.file_name()?.to_str()?; // relies on UTF-8 file name
    let parent = target.parent().unwrap_or_else(|| Path::new("."));
    let prefix = format!(".{name}.{tag}.");

    // Track best (timestamp, base_path) as we scan â€” no Vec, no sort.
    let mut best: Option<(u128, PathBuf)> = None;

    let rd = fs::read_dir(parent).ok()?;
    for entry in rd.flatten() {
        let fname = entry.file_name(); // OsString
        let Some(s) = fname.to_str() else { continue }; // skip non-UTF-8

        // Must start with the prefix
        let Some(rest) = s.strip_prefix(&prefix) else {
            continue;
        };

        // Accept ".bak" or ".bak.meta.json"
        let Some(num_s) = rest
            .strip_suffix(".bak")
            .or_else(|| rest.strip_suffix(".bak.meta.json"))
        else {
            continue;
        };

        let Ok(ts) = num_s.parse::<u128>() else {
            continue;
        };

        // If this timestamp is newer, keep it
        let is_better = best.as_ref().is_none_or(|(cur, _)| ts > *cur);
        if is_better {
            // Construct the .bak base path (sidecar is derived later)
            let base = parent.join(format!("{prefix}{ts}.bak"));
            best = Some((ts, base));
        }
    }

    let (_, base) = best?;
    let sidecar = sidecar_path_for_backup(&base);
    let backup_present = if base.exists() { Some(base) } else { None };
    Some((backup_present, sidecar))
}

/// Return the previous (second newest) backup pair if present.
pub(crate) fn find_previous_backup_and_sidecar(
    target: &Path,
    tag: &str,
) -> Option<(Option<PathBuf>, PathBuf)> {
    let name = target.file_name()?.to_str()?;
    let parent = target.parent().unwrap_or_else(|| Path::new("."));
    let prefix = format!(".{name}.{tag}.");

    let mut seen = HashSet::<u128>::new();

    // Collect unique (timestamp, base_path) pairs
    let mut stamps: Vec<(u128, PathBuf)> = fs::read_dir(parent)
        .ok()?
        .filter_map(Result::ok)
        .filter_map(|e| e.file_name().into_string().ok())
        .filter_map(|s| {
            // Guard: must start with prefix
            let rest = s.strip_prefix(&prefix)?;
            // Accept either ".bak" or ".bak.meta.json"
            let num_s = rest
                .strip_suffix(".bak")
                .or_else(|| rest.strip_suffix(".bak.meta.json"))?;
            let num: u128 = num_s.parse().ok()?;
            // Deduplicate timestamps
            if !seen.insert(num) {
                return None;
            }
            let base = parent.join(format!("{prefix}{num}.bak"));
            Some((num, base))
        })
        .collect();

    if stamps.len() < 2 {
        return None;
    }

    // Second newest by timestamp
    stamps.sort_unstable_by_key(|(ts, _)| *ts);
    let (_ts, base) = stamps.get(stamps.len() - 2)?.clone();

    let sidecar = sidecar_path_for_backup(&base);
    let backup_present = if base.exists() { Some(base) } else { None };

    Some((backup_present, sidecar))
}
