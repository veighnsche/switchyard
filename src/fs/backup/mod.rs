//! Backup subsystem â€” idiomatic directory module

pub mod snapshot;
pub mod sidecar;
pub mod index;

pub use snapshot::*;
pub use sidecar::*;
pub use index::*;
