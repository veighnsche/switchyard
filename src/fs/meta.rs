//! Filesystem metadata helpers used by preflight/apply for Minimal Facts and gating.
//!
//! This module provides conservative, non-mutating probes for:
//! - `kind_of(path)`: classify node kind (file/dir/symlink/missing/unknown)
//! - `resolve_symlink_target(path)`: resolve symlink target to an absolute path
//! - `detect_preservation_capabilities(path)`: detect which preservation dimensions are likely
//!   supported on the current platform and under current privileges.
//!
//! Notes:
//! - Owner preservation is reported true only when the effective UID is 0 (root).
//! - xattrs support is probed via the `xattr` crate by attempting to list attributes.
//! - Timestamps and mode are reported true when metadata is readable for the path.
//! - ACLs and capabilities are conservatively reported false.
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};

use serde_json::json;

/// Compute SHA-256 of a file at `path`, returning a lowercase hex string.
pub fn sha256_hex_of(path: &Path) -> Option<String> {
    let mut f = std::fs::File::open(path).ok()?;
    let mut hasher = Sha256::new();
    std::io::copy(&mut f, &mut hasher).ok()?;
    let out = hasher.finalize();
    Some(hex::encode(out))
}

/// If `target` is a symlink, resolve its target to an absolute path.
/// Relative links are resolved relative to the parent directory of `target`.
pub fn resolve_symlink_target(target: &Path) -> Option<PathBuf> {
    if let Ok(md) = std::fs::symlink_metadata(target) {
        if md.file_type().is_symlink() {
            if let Ok(mut link) = std::fs::read_link(target) {
                if link.is_relative() {
                    if let Some(parent) = target.parent() {
                        link = parent.join(link);
                    }
                }
                return Some(link);
            }
        }
    }
    None
}

/// Return a string describing the kind of filesystem node at `path`.
pub fn kind_of(path: &Path) -> String {
    match std::fs::symlink_metadata(path) {
        Ok(md) => {
            let ft = md.file_type();
            if ft.is_symlink() {
                "symlink".to_string()
            } else if ft.is_file() {
                "file".to_string()
            } else if ft.is_dir() {
                "dir".to_string()
            } else {
                "unknown".to_string()
            }
        }
        Err(_) => "missing".to_string(),
    }
}

/// Heuristic preservation capability detector for target path.
/// Returns (preservation map, preservation_supported flag).
/// This is intentionally conservative and non-mutating; it checks basic platform support
/// and permission surface for:
/// - owner (chown on non-root will likely fail; we report false unless running as root)
/// - mode (chmod generally possible if we can access the file)
/// - timestamps (utimensat typically available; report true if file exists)
/// - xattrs (Linux extended attributes via getxattr syscall presence; best-effort probe)
/// - acls (no portable check; report false)
/// - caps (Linux file capabilities; report false unless libcap is present â€” we report false)
pub fn detect_preservation_capabilities(path: &Path) -> (serde_json::Value, bool) {
    let mut owner = false;
    let mut mode = false;
    let mut timestamps = false;
    let mut xattrs = false;
    let acls = false;
    let caps = false;

    if let Ok(md) = std::fs::symlink_metadata(path) {
        // mode: we can generally preserve permissions if file exists and we have write on parent
        mode = true;
        // timestamps: if metadata is readable, assume we can set atime/mtime (utimensat)
        timestamps = true;
        // owner: only root can chown arbitrarily; detect effective uid == 0 via /proc/self/status
        owner = effective_uid_is_root();
        // xattrs: probe availability by attempting to list extended attributes
        xattrs = xattr::list(path).map(|_| true).unwrap_or(false);
        let _ = md; // silence unused on non-unix targets
    }

    let preservation = json!({
        "owner": owner,
        "mode": mode,
        "timestamps": timestamps,
        "xattrs": xattrs,
        "acls": acls,
        "caps": caps,
    });
    // preservation_supported if any dimension can be preserved
    let supported = owner || mode || timestamps || xattrs || acls || caps;
    (preservation, supported)
}

fn effective_uid_is_root() -> bool {
    // Parse /proc/self/status and read the second field of the `Uid:` line (effective UID).
    // Fallback to false on any error for conservatism.
    if let Ok(mut f) = std::fs::File::open("/proc/self/status") {
        use std::io::Read as _;
        let mut s = String::new();
        if f.read_to_string(&mut s).is_ok() {
            for line in s.lines() {
                if line.starts_with("Uid:") {
                    // Format: Uid:\t<real>\t<eff>\t<saved>\t<fs>
                    let parts: Vec<&str> = line.split_whitespace().collect();
                    if parts.len() >= 3 {
                        if let Ok(eff) = parts[2].parse::<u32>() {
                            return eff == 0;
                        }
                    }
                }
            }
        }
    }
    false
}
