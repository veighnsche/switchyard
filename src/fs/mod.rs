//! Filesystem primitives used by Switchyard.
//!
//! This module provides low-level, TOCTOU-safe filesystem operations used by the
//! higher-level API stages. Consumers should prefer calling public API stages in
//! `switchyard::api` rather than these atoms. Low-level atoms are crate-private
//! and are not re-exported at the module root.

pub mod atomic;
pub mod backup;
pub mod meta;
pub mod mount;
pub mod paths;
pub mod restore;
pub mod swap;
