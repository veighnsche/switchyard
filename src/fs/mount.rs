//! Filesystem mount inspection and policy helpers.

use crate::types::{MountError, MountFlags};
use std::path::{Path, PathBuf};

pub trait MountInspector {
    /// Get mount flags for a path.
    ///
    /// # Errors
    ///
    /// Returns a `MountError` if mount information cannot be determined.
    fn flags_for(&self, path: &Path) -> Result<MountFlags, MountError>;
}

/// Production inspector. Prefer kernel syscalls when available; fall back to parsing /proc/self/mounts.
#[derive(Debug, Copy, Clone)]
pub struct ProcStatfsInspector;

impl ProcStatfsInspector {
    fn parse_proc_mounts(path: &Path) -> Result<MountFlags, MountError> {
        // Canonicalize best-effort; if it fails, still proceed with the raw path
        let p = path.canonicalize().unwrap_or_else(|_| path.to_path_buf());
        let content =
            std::fs::read_to_string("/proc/self/mounts").map_err(|_| MountError::Unknown)?;
        let mut best: Option<(PathBuf, String)> = None;
        for line in content.lines() {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() < 4 {
                continue;
            }
            let mnt = parts.get(1).map(PathBuf::from).ok_or(MountError::Unknown)?;
            if p.starts_with(&mnt) {
                let opts = parts
                    .get(3)
                    .ok_or(MountError::Unknown)?
                    .to_ascii_lowercase();
                match &best {
                    None => best = Some((mnt, opts)),
                    Some((b, _)) => {
                        if mnt.as_os_str().len() > b.as_os_str().len() {
                            best = Some((mnt, opts));
                        }
                    }
                }
            }
        }
        if let Some((_mnt, opts)) = best {
            let has_rw = opts.split(',').any(|o| o == "rw");
            let noexec = opts.split(',').any(|o| o == "noexec");
            Ok(MountFlags {
                read_only: !has_rw,
                no_exec: noexec,
            })
        } else {
            Err(MountError::Unknown)
        }
    }
}

impl MountInspector for ProcStatfsInspector {
    fn flags_for(&self, path: &Path) -> Result<MountFlags, MountError> {
        // For now, rely on /proc/self/mounts. A future improvement can add rustix::fs::statfs mapping.
        Self::parse_proc_mounts(path)
    }
}

/// Policy helper: Ensure that the filesystem containing `path` is mounted with both read and write permissions.
///
/// # Errors
///
/// Returns a `MountError` if the filesystem is not mounted with read and write permissions.
pub fn ensure_rw_exec(inspector: &impl MountInspector, path: &Path) -> Result<(), MountError> {
    match inspector.flags_for(path) {
        Ok(flags) => {
            if flags.read_only || flags.no_exec {
                return Err(MountError::Unknown);
            }
            Ok(())
        }
        Err(e) => Err(e),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct MockInspector {
        flags: Result<MountFlags, MountError>,
    }
    impl MountInspector for MockInspector {
        fn flags_for(&self, _path: &Path) -> Result<MountFlags, MountError> {
            self.flags.clone()
        }
    }

    #[test]
    fn ensure_rw_exec_passes_on_rw_exec() {
        let ins = MockInspector {
            flags: Ok(MountFlags {
                read_only: false,
                no_exec: false,
            }),
        };
        assert!(ensure_rw_exec(&ins, Path::new("/tmp")).is_ok());
    }

    #[test]
    fn ensure_rw_exec_fails_on_ro_or_noexec() {
        let ins1 = MockInspector {
            flags: Ok(MountFlags {
                read_only: true,
                no_exec: false,
            }),
        };
        assert!(ensure_rw_exec(&ins1, Path::new("/tmp")).is_err());
        let ins2 = MockInspector {
            flags: Ok(MountFlags {
                read_only: false,
                no_exec: true,
            }),
        };
        assert!(ensure_rw_exec(&ins2, Path::new("/tmp")).is_err());
    }

    #[test]
    fn ensure_rw_exec_fails_on_ambiguous() {
        let ins = MockInspector {
            flags: Err(MountError::Unknown),
        };
        assert!(ensure_rw_exec(&ins, Path::new("/tmp")).is_err());
    }
}
