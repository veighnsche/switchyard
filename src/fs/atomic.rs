//! Atomic symlink swap primitives and helpers.
//!
//! This module implements a TOCTOU-safe sequence using directory handles:
//! `open_dir_nofollow(parent) -> symlinkat(tmp) -> renameat(tmp, final) -> fsync(dirfd)`.
//!
//! Test override knobs:
//! - `SWITCHYARD_FORCE_EXDEV=1` â€” simulate a cross-filesystem rename error (EXDEV) to exercise
//!   degraded fallback paths and telemetry in higher layers.
use std::fs;
use std::path::Path;

use crate::constants::TMP_SUFFIX;
use rustix::fd::OwnedFd;
use rustix::fs::{openat, renameat, symlinkat, unlinkat, AtFlags, Mode, OFlags, CWD};
use rustix::io::Errno;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Instant;

fn errno_to_io(e: Errno) -> std::io::Error {
    std::io::Error::from_raw_os_error(e.raw_os_error())
}

// Global counter to produce unique temporary names within a process.
static NEXT_TMP_COUNTER: AtomicU64 = AtomicU64::new(0);

/// Open a directory with `O_DIRECTORY` | `O_NOFOLLOW` for atomic operations.
///
/// # Errors
///
/// Returns an IO error if the directory cannot be opened.
pub fn open_dir_nofollow(dir: &Path) -> std::io::Result<OwnedFd> {
    use std::os::unix::ffi::OsStrExt;
    let c = std::ffi::CString::new(dir.as_os_str().as_bytes())
        .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid path"))?;
    openat(
        CWD,
        c.as_c_str(),
        OFlags::RDONLY | OFlags::DIRECTORY | OFlags::CLOEXEC | OFlags::NOFOLLOW,
        Mode::empty(),
    )
    .map_err(errno_to_io)
}

/// Fsync the parent directory of `path` for durability.
///
/// # Errors
///
/// Returns an IO error if the parent directory cannot be opened or fsynced.
pub fn fsync_parent_dir(path: &Path) -> std::io::Result<()> {
    if let Some(parent) = path.parent() {
        let dir = fs::File::open(parent)?;
        dir.sync_all()?;
    }
    Ok(())
}

/// Fsync a directory using an already-open directory file descriptor.
///
/// This avoids a TOCTOU window from re-opening the directory by path.
fn fsync_dirfd(dirfd: &OwnedFd) -> std::io::Result<()> {
    rustix::fs::fsync(dirfd).map_err(errno_to_io)
}

/// Atomically swap a symlink target using a temporary file and renameat.
///
/// # Errors
///
/// Returns an IO error if the atomic swap operation fails.
pub fn atomic_symlink_swap(
    source: &Path,
    target: &Path,
    allow_degraded: bool,
    force_exdev: Option<bool>,
) -> std::io::Result<(bool, u64)> {
    use std::os::unix::ffi::OsStrExt;
    // Open parent directory with O_DIRECTORY | O_NOFOLLOW to prevent traversal and races
    let parent = target.parent().unwrap_or_else(|| Path::new("."));
    let fname = target
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or("target");
    let pid = std::process::id();
    let ctr = NEXT_TMP_COUNTER.fetch_add(1, Ordering::Relaxed);
    let tmp_name = format!(".{fname}.{pid}.{ctr}{TMP_SUFFIX}");

    let dirfd = open_dir_nofollow(parent)?;

    // Best-effort unlink temporary name if present (ignore ENOENT only)
    let tmp_c = std::ffi::CString::new(tmp_name.as_str())
        .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid cstring"))?;
    match unlinkat(&dirfd, tmp_c.as_c_str(), AtFlags::empty()) {
        Ok(()) => {}
        Err(e) if e == Errno::NOENT => {}
        Err(e) => return Err(errno_to_io(e)),
    }

    // Create symlink using symlinkat relative to parent dirfd
    let src_c = std::ffi::CString::new(source.as_os_str().as_bytes()).map_err(|_| {
        std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid source path")
    })?;
    let tmp_c2 = std::ffi::CString::new(tmp_name.as_str())
        .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid cstring"))?;
    symlinkat(src_c.as_c_str(), &dirfd, tmp_c2.as_c_str()).map_err(errno_to_io)?;

    // Atomically rename tmp -> final name within the same directory (bytes-safe)
    let new_c = if let Some(name_os) = target.file_name() {
        std::ffi::CString::new(name_os.as_bytes())
            .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid cstring"))?
    } else {
        std::ffi::CString::new("target")
            .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid cstring"))?
    };
    let t0 = Instant::now();
    let rename_res = renameat(&dirfd, tmp_c2.as_c_str(), &dirfd, new_c.as_c_str());
    // Prefer per-instance override when provided; fallback to gated env for legacy tests only.
    // Note: do not enable env overrides by default in tests to avoid cross-scenario interference.
    let allow_env_overrides =
        std::env::var_os("SWITCHYARD_TEST_ALLOW_ENV_OVERRIDES") == Some(std::ffi::OsString::from("1"));
    let inject_exdev = match force_exdev {
        Some(true) => true,
        Some(false) => false,
        None => {
            allow_env_overrides
                && std::env::var_os("SWITCHYARD_FORCE_EXDEV") == Some(std::ffi::OsString::from("1"))
        }
    };
    let rename_res = if inject_exdev {
        match rename_res {
            Ok(()) => Err(Errno::XDEV),
            Err(e) => Err(e),
        }
    } else {
        rename_res
    };
    match rename_res {
        Ok(()) => {
            let _ = fsync_dirfd(&dirfd);
            let fsync_ms = u64::try_from(t0.elapsed().as_millis()).unwrap_or(u64::MAX);
            Ok((false, fsync_ms))
        }
        Err(e) if e == Errno::XDEV && allow_degraded => {
            // Fall back: best-effort non-atomic replacement
            match unlinkat(&dirfd, new_c.as_c_str(), AtFlags::empty()) {
                Ok(()) => {}
                Err(e) if e == Errno::NOENT => {}
                Err(e) => return Err(errno_to_io(e)),
            }
            symlinkat(src_c.as_c_str(), &dirfd, new_c.as_c_str()).map_err(errno_to_io)?;
            let _ = fsync_dirfd(&dirfd);
            let fsync_ms = u64::try_from(t0.elapsed().as_millis()).unwrap_or(u64::MAX);
            Ok((true, fsync_ms))
        }
        Err(e) => Err(errno_to_io(e)),
    }
}
