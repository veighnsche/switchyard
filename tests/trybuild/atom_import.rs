// This test intentionally tries to import a crate-private FS atom.
// It should fail to compile.
use switchyard::fs::atomic_symlink_swap;

fn main() {
    let _ = atomic_symlink_swap;
}
