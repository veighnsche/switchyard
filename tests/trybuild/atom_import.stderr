error[E0432]: unresolved import `switchyard::fs::atomic_symlink_swap`
 --> tests/trybuild/atom_import.rs:3:5
  |
3 | use switchyard::fs::atomic_symlink_swap;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `atomic_symlink_swap` in `fs`
